bundles/org.eclipse.e4.ui.workbench/src/org/eclipse/e4/ui/internal/workbench/PartServiceImpl.java

```diff
@@ -727,12 +745,23 @@ private void activate(MPart part, boolean requiresFocus, boolean activateBranch)
 		IEclipseContext windowContext = window.getContext();
 		// check if the active part has changed or if we are no longer the active window
 		if (windowContext.getParent().getActiveChild() == windowContext && part == activePart) {
-			// insert it in the beginning of the activation history, it may not have been inserted
-			// pending when this service was instantiated
-			partActivationHistory.prepend(part);
-			UIEvents.publishEvent(UIEvents.UILifeCycle.ACTIVATE, part);
-			if (Policy.DEBUG_FOCUS) {
-				Activator.trace(Policy.DEBUG_FOCUS_FLAG, "Trying to activate already active part: " + part, null);//$NON-NLS-1$
+			if (!focusingActivePart) {
```

Comment:
This is getting ugly... We can also take code from `PartRenderingEngine.focusGui(MUIElement)` to only try to force focus by calling `AbstractPartRenderer.forceFocus(MUIElement)`... if that works differently. But I'm not sure that works any better.

What about just leaving the double focusing in? Maybe it causes trouble in some client code which starts receiving 2 focus events instead of 0... With the flag for switching perspective protecting the focus call, I think that should be the only case of different behavior.

Reply:
Let say it this way: most of e4 is ugly, but what should we do? At least this works.